pkg/controller/sriovnetwork/sriovnetwork_controller.go

```diff
@@ -143,35 +126,49 @@ func (r *ReconcileSriovNetwork) Reconcile(request reconcile.Request) (reconcile.
 	if err != nil {
 		return reconcile.Result{}, err
 	}
-
 	// Set SriovNetwork instance as the owner and controller
 	if err := controllerutil.SetControllerReference(instance, netAttDef, r.scheme); err != nil {
 		return reconcile.Result{}, err
 	}
-	
-	// Try initialize lastNetworkNamespace
+
+	// Check if this NetworkAttachmentDefinition already exists
+	found := &netattdefv1.NetworkAttachmentDefinition{}
+	err = r.client.Get(context.TODO(), types.NamespacedName{Name: netAttDef.Name, Namespace: netAttDef.Namespace}, found)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			reqLogger.Info("NetworkAttachmentDefinition not exist, creating")
+			err = r.client.Create(context.TODO(), netAttDef)
+			if err != nil {
+				reqLogger.Error(err, "Couldn't create NetworkAttachmentDefinition", "Namespace", netAttDef.Namespace, "Name", netAttDef.Name)
+				return reconcile.Result{}, err
+			}
+		}
+		reqLogger.Error(err, "Couldn't get NetworkAttachmentDefinition", "Namespace", netAttDef.Namespace, "Name", netAttDef.Name)
+		return reconcile.Result{}, err	
+	} else {
+		reqLogger.Info("NetworkAttachmentDefinition already exist, updating")
+		netAttDef.SetResourceVersion(found.GetResourceVersion())
+		err = r.client.Update(context.TODO(), netAttDef)
+		if err != nil {
+			reqLogger.Error(err, "Couldn't update NetworkAttachmentDefinition", "Namespace", netAttDef.Namespace, "Name", netAttDef.Name)
+			return reconcile.Result{}, err
+		}
+	}
+
+	// Check if there are more than one children for one SriovNetwork CR.
 	nadList := &netattdefv1.NetworkAttachmentDefinitionList{}
-	fieldSelector := fields.Set{"metadata.name": instance.GetName(),}.AsSelector()
-	r.client.List(context.TODO(), &client.ListOptions{FieldSelector: fieldSelector,}, nadList)
+	r.client.List(context.TODO(), &client.ListOptions{}, nadList)
 	sriovNads := []netattdefv1.NetworkAttachmentDefinition{}
 	for _, cr := range nadList.Items {
 		if cr.GetOwnerReferences()[0].UID == instance.UID {
 			sriovNads = append(sriovNads, cr)
 		}
 	}
-	reqLogger.Info("NetworkAttachmentDefinition", "list", nadList)
-	if len(sriovNads) == 1 {
-		lastNetworkNamespace[instance.GetUID()] = sriovNads[0].GetNamespace()
-	} else if len(sriovNads) > 1 {
+	reqLogger.Info("NetworkAttachmentDefinition", "list", sriovNads)
+	if len(sriovNads) > 1 {
 		reqLogger.Info("more than one NetworkAttachmentDefinition CR exists for one SriovNetwork CR", "Namespace", instance.GetNamespace(), "Name", instance.GetName())
-		ns := ""
-		if  instance.Spec.NetworkNamespace != "" {
-			ns = instance.Spec.NetworkNamespace
-		} else {
-			ns = instance.GetNamespace()
-		}
 		for _, nad := range sriovNads {
-			if nad.GetNamespace() != ns {
+			if nad.GetNamespace() != netAttDef.GetNamespace() {
```

Comment:
This may wrongly delete net-attach-def CRs that have same name as defined in SriovNetwork, but in a different namespace. Shall we check CNI type here and only apply when it's `sriov`?
CNO may support creating net-attach-def for other types of CNI. If both try to recover the deleted net-attach-def, will there be confliction between CNO and SRIOV Operator?

Reply:
It will not, there is a check above to only put CR with matched owner ID (which is the ID of the SriovNetwork CR) to `sriovNads`.

Reply:
Yes, I didn't notice there is a loop to filter net-attach-def created by sriovnetwork.